File: LegacySource/settlement/room/main/placement/UtilStructure.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Settlement.Room.Main.Placement
{
    public sealed class UtilStructure : GETTER_IMP<TBuilding>
    {
        private readonly RoomPlacer p;
        private readonly PropGame saver = PROP.Game("ROOM_CONSTRUCTION");

        public UtilStructure(RoomPlacer p)
        {
            this.p = p;

            Set(SETT.TERRAIN().BUILDINGS.MUD);
        }

        public SETT_JOB GetWallJob(int tx, int ty)
        {
            if (SETT.TERRAIN().CAVE.Is(tx, ty))
                return SETT.JOBS().clearss.caveFill;
            return SETT.JOBS().build_structure.Get(Get().structure.Index()).wall;
        }

        public SETT_JOB GetCeilingJob(int tx, int ty)
        {
            if (SETT.TERRAIN().MOUNTAIN.Is(tx, ty))
                return null;
            if (Get().roof.Is(tx, ty))
                return null;
            if (SETT.TERRAIN().CAVE.Is(tx, ty))
                return null;
            return SETT.JOBS().build_structure.Get(Get().structure.Index()).ceiling;
        }

        public void Read()
        {
            string k = saver.Chars("STRUCTURE");
            if (k != null && STRUCTURES.Map().TryGet(k) != null)
            {
                Set(STRUCTURES.Map().TryGet(k).Terrain());
            }
        }

        private int unroofed = 0;
        private int tick = 0;

        public int Roofs()
        {
            if (GAME.UpdateI() != tick)
            {
                unroofed = 0;
                foreach (COORDINATE c in p.instance.Body())
                {
                    if (!p.instance.Is(c))
                        continue;

                    if (GetCeilingJob(c.X(), c.Y()) != null)
                    {
                        unroofed++;
                    }
                }
                if (p.autoWalls.Is())
                {
                    unroofed += p.door.GetOpenings();
                }
            }
            return unroofed;
        }

        public override void Set(TBuilding t)
        {
            base.Set(t);
            saver.CharsSet("STRUCTURE", t.structure.key);
        }

        public int Walls()
        {
            if (!p.autoWalls.Is())
                return 0;
            return p.door.GetWalls();
        }

        public int MountainWalls()
        {
            if (!p.autoWalls.Is())
                return 0;
            return p.door.GetMountains();
        }

        public void Set(int tx, int ty)
        {
            TerrainTile t = SETT.TERRAIN().Get(tx, ty);
            if (t is TBuilding.BuildingComponent)
            {
                Set(((TBuilding.BuildingComponent)t).Building());
            }
        }
    }
}

File: settlement/room/main/placement/UtilStructure.java
package settlement.room.main.placement;

import game.GAME;
import game.save.PROP;
import game.save.PROP.PropGame;
import init.structure.STRUCTURES;
import settlement.main.SETT;
import settlement.misc.job.SETT_JOB;
import settlement.tilemap.terrain.TBuilding;
import settlement.tilemap.terrain.Terrain.TerrainTile;
import snake2d.util.datatypes.COORDINATE;
import util.data.GETTER.GETTER_IMP;

public final class UtilStructure extends GETTER_IMP<TBuilding> {
	
	private final RoomPlacer p;
	private final PropGame saver = PROP.game("ROOM_CONSTRUCTION");
	
	UtilStructure(RoomPlacer p){
		this.p = p;
		
		set(SETT.TERRAIN().BUILDINGS.MUD);
		
		
	}
	
	SETT_JOB getWallJob(int tx, int ty) {
		
		if (SETT.TERRAIN().CAVE.is(tx, ty))
			return SETT.JOBS().clearss.caveFill;
		return SETT.JOBS().build_structure.get(get().structure.index()).wall;
		
	}
	
	SETT_JOB getCeilingJob(int tx, int ty) {
		
		if (SETT.TERRAIN().MOUNTAIN.is(tx, ty))
			return null;
		if (get().roof.is(tx, ty))
			return null;
		if (SETT.TERRAIN().CAVE.is(tx, ty))
			return null;
		return SETT.JOBS().build_structure.get(get().structure.index()).ceiling;
	}
	
	void read() {
		String k = saver.chars("STRUCTURE");
		if (k != null && STRUCTURES.map().tryGet(k) != null) {
			set(STRUCTURES.map().tryGet(k).terrain());
		}
		
	}
	
	int unroofed = 0;
	int tick = 0;
	
	public int roofs() {
		
		if (GAME.updateI() != tick) {
			unroofed = 0;
			for (COORDINATE c : p.instance.body()) {
				if (!p.instance.is(c))
					continue;
				
				if (getCeilingJob(c.x(), c.y()) != null) {
					unroofed ++;
				}
			}
			if (p.autoWalls.is()) {
				unroofed += p.door.getOpenings();
			}
		}
		return unroofed;
	}
	@Override
	public void set(TBuilding t) {
		super.set(t);
		saver.charsSet("STRUCTURE", t.structure.key);
	}
	public int walls() {
		if (!p.autoWalls.is())
			return 0;
		return p.door.getWalls();
	}
	
	public int mountainWalls() {
		if (!p.autoWalls.is())
			return 0;
		return p.door.getMountains();
	}

	public void set(int tx, int ty) {
		TerrainTile t = SETT.TERRAIN().get(tx, ty);
		if (t instanceof TBuilding.BuildingComponent) {
			set(((TBuilding.BuildingComponent)t).building());
		}
			
		
	}
	
	
	
}
